docs/generatedcodeguide.rst

```diff
@@ -0,0 +1,228 @@
+    // to the generated message byte stream.
+    var unknownFieldStream: bytes = "";
+    
+    // User exposed method for serializing data to protobuf wire format.
+    // This is a wrapper method to the actual method, it is used for making a
+    // binary writing channel out of the channel given as an input by the user. 
+    proc serialize(ch) throws { ... }
+    
+    // Contains the actual implementation for serializing data. Calls the
+    // append functions of the user support library. It appends the `unknownFieldStream`
+    // at the end of the message.
+    proc _serialize(binCh) throws { ... }
+    
+    // User exposed method for parsing data from protobuf wire format.
+    // This is a wrapper method to the actual method, it is used for making a
+    // binary reading channel out of the channel given as an input by the user. 
+    proc deserialize(ch) throws { ... }
+    
+    // Contains the actual implementation for parsing data. Calls the consume
+    // functions of the user support library. Appends unknown fields encountered
+    // to the `unknownFieldStream` variable.
+    proc _deserialize(binCh) throws { ... }
+  
+  }  
+
+Nested Types
+------------
+A message can be declared inside another message. For example:
+
+.. code-block:: proto
+
+  message Foo {
+    message Bar {
+      ...
+    }
+  }
+
+In this case, or if a message contains a nested enum, the compiler will generate
+module level records/enums with a name prefixed by the parent message name:
+
+.. code-block:: chpl
+  
+  record Foo {
+   ...
+  }
+  
+  // Nested Types
+  record Foo_Bar {
+    ...
+  }
+
+.. note::
+  Nested records or declaration of enums in records is currently not supported in
+  Chapel. Once we have support for these, we can declare nested type in the parent
+  record and thus avoiding the name prefix. 
+
+Fields
+------
+The protocol buffer compiler generates a Chapel record field for each field defined
+within a message. In case of Chapel, methods equivalent to `get` and `set` in other
+languages are implicitly generated by the Chapel compiler.
+
+Scalar Value Types
+^^^^^^^^^^^^^^^^^^
+A scalar message field can have one of the following types, the table shows the
+type specified in the ``.proto`` file, and the corresponding generated Chapel type:
+
+.. list-table::
+   :widths: 50 50
+   :header-rows: 1
+
+   * - .proto Type
+     - Chapel Type
+   * - double
+     - real(64)
+   * - float
+     - real(32)
+   * - int32
+     - int(32)
+   * - int64
+     - int(64)
+   * - uint32
+     - uint(32)
+   * - uint64
+     - uint(64)
+   * - sint32
+     - int(32)
+   * - sint64
+     - int(64)
+   * - fixed32
+     - uint(32)
+   * - fixed64
+     - uint(64)
+   * - sfixed32
+     - int(32)
+   * - sfixed64
+     - int(64)
+   * - bool
+     - bool
+   * - string
+     - string
+   * - bytes
+     - bytes
+
+Singular Fields
+^^^^^^^^^^^^^^^
+Every singular field generates a record field variable of an appropriate Chapel type. 
```

Comment:
It might be helpful to link to [this section](https://developers.google.com/protocol-buffers/docs/proto3#specifying_field_rules) here, in case users forgot about it when reading the documentation (or were lazy and skipped it ;) )